### ISSUE TITLE
DataStore codegen changes for upcoming iOS release

### ISSUE THREAD

**Issue by drochetti** [aws-amplify/amplify-cli#3506]
### Description

When generating code for GraphQL schemas on iOS, some features are missing, such as enum and non-model types. This ticket highlights the required changes for the upcoming iOS release of Amplify DataStore.

#### Enums:

Given the GraphQL schema:

```graphql
enum PostStatus {
    active
    inactive
    draft
}
```

The expected iOS code is:

```swift
import Amplify

public enum PostStatus: String, EnumPersistable {

    case active
    case inactive
    case draft

}
```

**Notes:**

- Enum values in Swift should be `camelCase`
- The protocol being added (`EnumPersistable`) 
- Some enum values might clash with reserved keywords in Swift, in that case it needs to be escaped. To escape in Swift use back-ticks: `` `public` ``
- it's **not necessary** to register Enums in the `ModelRegistry`

#### Non-model types (aka Nested or Embedded types)

For non-model types, the codegen should generate Swift structs that follow the same semantic of the model types minus the Model-specific rules such as `: Model` and using `List<M>` for associations (since associations are not supported in non-model types).

```graphql
type Location {
    latitude: Float!
    longitude: Float!
    tags: [String]
}
```

The expected iOS code is:

```swift
public struct Location: Codable {

    var latitude: Double
    var longitude: Double
    var tags: [String]

}
```

**Notes:**

- reuse all the types from models but:
  - doesn't conform to `Model`
  - conforms to `Codable`
- never use `List<M>`
- models cannot be used as property types of non-model types (I believe this is enforced at the transformer level already, is that correct?)

#### Schema definition

On the schema definition (e.g. `Post+Schema.swift`) when dealing with non-model types **and arrays of non-model/scalar types** the definition should be like:

```swift
.customType(Location.self)
.customType([String].self)

// complete definition of a field:

.field(example.nonModelField, is: .required, ofType: .customType(ExampleNonModelType.self))

// .collection(of) is only used for Model types
.field(example.arrayOfStringsField, is: .required, ofType: .customType([String].self))
```

### Related issues:

- https://github.com/aws-amplify/amplify-ios/issues/240
  - https://github.com/aws-amplify/amplify-ios/issues/246
  - https://github.com/aws-amplify/amplify-ios/issues/257
- https://github.com/aws-amplify/amplify-ios/issues/111

**Edit:** added information about how to define the schema for the non-model types

**Edit 2:**
- final name for the enum protocol: `EnumPersistable`
- final name for non-model types and arrays of scalars/types: `.customType` (code samples above updated accordingly)

**Comment by drochetti** [comment]
The iOS PR https://github.com/aws-amplify/amplify-ios/pull/334 landed in master. The names were updated in the original issue and reflect the final names.

**Comment by yuth** [comment]
Change have been released in Amplify CLI v 4.17.2
